File: web/src/lib/push.ts
/** Push notifications with tap-to-pick buttons, via ntfy.sh.
 *
 * ntfy gives real actionable notifications (HTTP action buttons) with no app to
 * build: install ntfy, subscribe to your private topic, and each offered slot
 * becomes a button that POSTs back to our /api/decide route. ntfy allows up to
 * 3 buttons, so we show up to 2 slots + Decline.
 */
import { config, requireConfig } from "./config";
import { CallRecord, OfferedSlot } from "./types";

const MAX_SLOT_BUTTONS = 2;

function decideUrl(callId: string, choice: string): string {
  const qs = new URLSearchParams({ call: callId, choice }).toString();
  return `${config.publicBaseUrl}/api/decide?${qs}`;
}

export async function sendDecisionRequest(
  record: CallRecord,
  slots: OfferedSlot[],
): Promise<void> {
  requireConfig("ntfyTopic", "publicBaseUrl");
  const shown = slots.slice(0, MAX_SLOT_BUTTONS);
  const actions = shown.map((slot, i) => ({
    action: "http",
    label: slot.startsAt.slice(0, 24),
    url: decideUrl(record.id, String(i)),
    method: "POST",
    clear: true,
  }));
  actions.push({
    action: "http",
    label: "Decline",
    url: decideUrl(record.id, "no"),
    method: "POST",
    clear: true,
  });

  // PHI-free push: ntfy is a third-party broker, so no patient name, provider,
  // or reason goes here. Just an action prompt + slot times (a bare time isn't
  // PHI). Open the app to see who/what it's for.
  await fetch(config.ntfyServer, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      topic: config.ntfyTopic,
      title: "Appointment approval needed",
      message: "An offered time is outside your usual window. Tap a time to book, or decline.",
      priority: 5,
      tags: ["calendar"],
      actions,
    }),
  });
}

export async function sendInfo(title: string, message: string): Promise<void> {
  if (!config.ntfyTopic) return;
  await fetch(config.ntfyServer, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ topic: config.ntfyTopic, title, message, tags: ["bell"] }),
  });
}

File: python/app/push.py
"""Push notifications with tap-to-pick buttons, via ntfy.sh.

ntfy gives us real actionable notifications (HTTP action buttons) with no app to
build: install the ntfy app, subscribe to your private topic, and each offered
slot becomes a button that POSTs straight back to our /decide endpoint.

ntfy allows up to 3 action buttons, so we show up to 2 slots + a Decline button.
If the office offered more, the agent is told to read back only its top 2.
"""
from __future__ import annotations

from typing import List
from urllib.parse import urlencode

import httpx

from .config import get_settings
from .models import CallRecord, OfferedSlot

_MAX_SLOT_BUTTONS = 2


def _decide_url(call_id: str, choice: str) -> str:
    s = get_settings()
    return f"{s.public_base_url}/decide?" + urlencode({"call": call_id, "choice": choice})


def send_decision_request(record: CallRecord, slots: List[OfferedSlot]) -> None:
    """Notify the owner with tap-to-pick buttons for the offered slots."""
    s = get_settings()
    s.require("ntfy_topic", "public_base_url")

    shown = slots[:_MAX_SLOT_BUTTONS]
    actions = []
    for i, slot in enumerate(shown):
        actions.append(
            {
                "action": "http",
                "label": slot.starts_at[:24],
                "url": _decide_url(record.id, str(i)),
                "method": "POST",
                "clear": True,
            }
        )
    actions.append(
        {
            "action": "http",
            "label": "Decline",
            "url": _decide_url(record.id, "no"),
            "method": "POST",
            "clear": True,
        }
    )

    req = record.request
    body = {
        "topic": s.ntfy_topic,
        "title": f"Approve appointment? {req.patient.name}",
        "message": (
            f"{req.provider_name} offered times outside your usual window "
            f"for: {req.reason}. Tap to pick or decline."
        ),
        "priority": 5,  # max — make the phone buzz now; the operator is holding
        "tags": ["calendar"],
        "actions": actions,
    }
    httpx.post(s.ntfy_server, json=body, timeout=15).raise_for_status()


def send_info(title: str, message: str) -> None:
    """Fire-and-forget informational push (booked / escalated / no slots)."""
    s = get_settings()
    if not s.ntfy_topic:
        return
    httpx.post(
        s.ntfy_server,
        json={"topic": s.ntfy_topic, "title": title, "message": message, "tags": ["bell"]},
        timeout=15,
    )
